File: bot-api/dotnet/src/BaseBot.cs
using System;
using System.Collections.Generic;
using Robocode.TankRoyale.BotApi.Events;
using Robocode.TankRoyale.BotApi.Internal;
using static System.Double;

namespace Robocode.TankRoyale.BotApi
{
  /// <summary>
  /// Abstract bot class that takes care of communication between the bot and the server, and sends
  /// notifications through the event handlers. Most bots can inherit from this class to get access
  /// to basic methods.
  /// </summary>
  public abstract class BaseBot : IBaseBot
  {
    internal readonly BaseBotInternals __baseBotInternals;

    /// <summary>
    /// Constructor for initializing a new instance of the BaseBot class.
    /// This constructor should be used when  both BotInfo and server URL is provided through
    /// environment variables, i.e., when starting up the bot using a booter. These environment
    /// variables must be set to provide the server URL and bot information, and are automatically
    /// set by the booter tool for Robocode.
    ///
    /// Example of how to set the predefined environment variables:
    ///
    /// SERVER_URL=ws://localhost<br/>
    /// SERVER_SECRET=xzoEeVbnBe5TGjCny0R1yQ<br/>
    /// BOT_NAME=MyBot<br/>
    /// BOT_VERSION=1.0<br/>
    /// BOT_AUTHORS=John Doe<br/>
    /// BOT_DESCRIPTION=A short description<br/>
    /// BOT_URL=http://somewhere.net/MyBot<br/>
    /// BOT_COUNTRY_CODES=us<br/>
    /// BOT_GAME_TYPES=classic, melee, 1v1<br/>
    /// BOT_PLATFORM=.Net 5.0<br/>
    /// BOT_PROG_LANG=C# 8.0<br/>
    /// </summary>
    public BaseBot()
    {
      __baseBotInternals = new BaseBotInternals(this, null, null, null);
    }

    /// <summary>
    /// Constructor for initializing a new instance of the BaseBot class.
    /// This constructor assumes the server URL and secret is provided by the environment
    /// variables SERVER_URL and SERVER_SECRET.
    /// </summary>
    /// <param name="botInfo">Is the bot info containing information about your bot.</param>
    public BaseBot(BotInfo botInfo)
    {
      __baseBotInternals = new BaseBotInternals(this, botInfo, null, null);
    }

    /// <summary>
    /// Constructor for initializing a new instance of the BaseBot class.
    /// </summary>
    /// <param name="botInfo">Is the bot info containing information about your bot.</param>
    /// <param name="serverUrl">Is the server URL</param>
    public BaseBot(BotInfo botInfo, Uri serverUrl)
    {
      __baseBotInternals = new BaseBotInternals(this, botInfo, serverUrl, null);
    }

    /// <summary>
    /// Constructor for initializing a new instance of the BaseBot class.
    /// </summary>
    /// <param name="botInfo">Is the bot info containing information about your bot.</param>
    /// <param name="serverUrl">Is the server URL</param>
    /// <param name="serverSecret">Is the server secret</param>
    public BaseBot(BotInfo botInfo, Uri serverUrl, string serverSecret)
    {
      __baseBotInternals = new BaseBotInternals(this, botInfo, serverUrl, serverSecret);
    }

    /// <inheritdoc/>
    public void Start()
    {
      __baseBotInternals.Start();
    }

    /// <inheritdoc/>
    public void Go()
    {
      __baseBotInternals.Execute();
    }

    /// <inheritdoc/>
    public string Variant => __baseBotInternals.Variant;

    /// <inheritdoc/>
    public string Version => __baseBotInternals.Version;

    /// <inheritdoc/>
    public int MyId => __baseBotInternals.MyId;

    /// <inheritdoc/>
    public string GameType => __baseBotInternals.GameSetup.GameType;

    /// <inheritdoc/>
    public int ArenaWidth => __baseBotInternals.GameSetup.ArenaWidth;

    /// <inheritdoc/>
    public int ArenaHeight => __baseBotInternals.GameSetup.ArenaHeight;

    /// <inheritdoc/>
    public int NumberOfRounds => __baseBotInternals.GameSetup.NumberOfRounds;

    /// <inheritdoc/>
    public double GunCoolingRate => __baseBotInternals.GameSetup.GunCoolingRate;

    /// <inheritdoc/>
    public int? MaxInactivityTurns => __baseBotInternals.GameSetup.MaxInactivityTurns;

    /// <inheritdoc/>
    public int TurnTimeout => __baseBotInternals.GameSetup.TurnTimeout;

    /// <inheritdoc/>
    public int TimeLeft => __baseBotInternals.TimeLeft;

    /// <inheritdoc/>
    public int RoundNumber => __baseBotInternals.CurrentTick.RoundNumber;

    /// <inheritdoc/>
    public int TurnNumber => __baseBotInternals.CurrentTick.TurnNumber;

    /// <inheritdoc/>
    public int EnemyCount => __baseBotInternals.CurrentTick.EnemyCount;

    /// <inheritdoc/>
    public double Energy => __baseBotInternals.CurrentTick.BotState.Energy;

    /// <inheritdoc/>
    public bool IsDisabled => Energy == 0;

    /// <inheritdoc/>
    public double X => __baseBotInternals.CurrentTick.BotState.X;

    /// <inheritdoc/>
    public double Y => __baseBotInternals.CurrentTick.BotState.Y;

    /// <inheritdoc/>
    public double Direction => __baseBotInternals.CurrentTick.BotState.Direction;

    /// <inheritdoc/>
    public double GunDirection => __baseBotInternals.CurrentTick.BotState.GunDirection;

    /// <inheritdoc/>
    public double RadarDirection => __baseBotInternals.CurrentTick.BotState.RadarDirection;

    /// <inheritdoc/>
    public double Speed => __baseBotInternals.CurrentTick.BotState.Speed;

    /// <inheritdoc/>
    public double GunHeat => __baseBotInternals.CurrentTick.BotState.GunHeat;

    /// <inheritdoc/>
    public IEnumerable<BulletState> BulletStates => __baseBotInternals.CurrentTick.BulletStates;

    /// <inheritdoc/>
    public IEnumerable<BotEvent> Events => __baseBotInternals.CurrentTick.Events;

    /// <inheritdoc/>
    public double TurnRate
    {
      set
      {
        if (IsNaN(value))
        {
          throw new ArgumentException("TurnRate cannot be NaN");
        }
        __baseBotInternals.BotIntent.TurnRate = value;
      }
      get => __baseBotInternals.CurrentTick.BotState.TurnRate;
    }

    /// <inheritdoc/>
    public void SetMaxTurnRate(double maxTurnRate)
    {
      __baseBotInternals.SetMaxTurnRate(maxTurnRate);
    }

    /// <inheritdoc/>
    public double GunTurnRate
    {
      set
      {
        if (IsNaN(value))
        {
          throw new ArgumentException("GunTurnRate cannot be NaN");
        }
        __baseBotInternals.BotIntent.GunTurnRate = value;
      }
      get => __baseBotInternals.CurrentTick.BotState.GunTurnRate;
    }

    /// <inheritdoc/>
    public void SetMaxGunTurnRate(double maxGunTurnRate)
    {
      __baseBotInternals.SetMaxGunTurnRate(maxGunTurnRate);
    }

    /// <inheritdoc/>
    public double RadarTurnRate
    {
      set
      {
        if (IsNaN(value))
        {
          throw new ArgumentException("RadarTurnRate cannot be NaN");
        }
        __baseBotInternals.BotIntent.RadarTurnRate = value;
      }
      get => __baseBotInternals.CurrentTick.BotState.RadarTurnRate;
    }

    /// <inheritdoc/>
    public void SetMaxRadarTurnRate(double maxRadarTurnRate)
    {
      __baseBotInternals.SetMaxRadarTurnRate(maxRadarTurnRate);
    }

    /// <inheritdoc/>
    public double TargetSpeed
    {
      set
      {
        if (IsNaN(value))
        {
          throw new ArgumentException("TargetSpeed cannot be NaN");
        }
        __baseBotInternals.BotIntent.TargetSpeed = value;
      }
      get => __baseBotInternals.BotIntent.TargetSpeed ?? 0d;
    }

    /// <inheritdoc/>
    public void SetMaxSpeed(double maxSpeed)
    {
      __baseBotInternals.SetMaxSpeed(maxSpeed);
    }

    /// <inheritdoc/>
    public bool SetFire(double firepower)
    {
      return __baseBotInternals.SetFire(firepower);
    }

    /// <inheritdoc/>
    public void SetScan()
    {
      __baseBotInternals.BotIntent.Scan = true;
    }

    /// <inheritdoc/>
    public double Firepower => __baseBotInternals.BotIntent.Firepower ?? 0;

    /// <inheritdoc/>
    public bool DoAdjustGunForBodyTurn
    {
      set => __baseBotInternals.BotIntent.AdjustGunForBodyTurn = value;
      get => __baseBotInternals.BotIntent.AdjustGunForBodyTurn ?? false;
    }

    /// <inheritdoc/>
    public bool DoAdjustRadarForGunTurn
    {
      set => __baseBotInternals.BotIntent.AdjustRadarForGunTurn = value;
      get => __baseBotInternals.BotIntent.AdjustRadarForGunTurn ?? false;
    }

    /// <inheritdoc/>
    public void AddCustomEvent(Condition condition)
    {
      __baseBotInternals.AddCondition(condition);
    }

    /// <inheritdoc/>
    public void RemoveCustomEvent(Condition condition)
    {
      __baseBotInternals.RemoveCondition(condition);
    }

    /// <inheritdoc/>
    public void SetStop()
    {
      __baseBotInternals.SetStop();
    }

    /// <inheritdoc/>
    public void SetResume()
    {
      __baseBotInternals.SetResume();
    }

    /// <inheritdoc/>
    public bool IsStopped => __baseBotInternals.IsStopped;

    /// <inheritdoc/>
    public int? GetBodyColor() => __baseBotInternals.CurrentTick.BotState.BodyColor;

    /// <inheritdoc/>
    public void SetBodyColor(string bodyColor) => __baseBotInternals.BotIntent.BodyColor = bodyColor;

    /// <inheritdoc/>
    public int? GetTurretColor() => __baseBotInternals.CurrentTick.BotState.TurretColor;

    /// <inheritdoc/>
    public void SetTurretColor(string turretColor) => __baseBotInternals.BotIntent.TurretColor = turretColor;

    /// <inheritdoc/>
    public int? GetRadarColor() => __baseBotInternals.CurrentTick.BotState.RadarColor;

    /// <inheritdoc/>
    public void SetRadarColor(string radarColor) => __baseBotInternals.BotIntent.RadarColor = radarColor;

    /// <inheritdoc/>
    public int? GetBulletColor() => __baseBotInternals.CurrentTick.BotState.BulletColor;

    /// <inheritdoc/>
    public void SetBulletColor(string bulletColor) => __baseBotInternals.BotIntent.BulletColor = bulletColor;

    /// <inheritdoc/>
    public int? GetScanColor() => __baseBotInternals.CurrentTick.BotState.ScanColor;

    /// <inheritdoc/>
    public void SetScanColor(string scanColor) => __baseBotInternals.BotIntent.ScanColor = scanColor;

    /// <inheritdoc/>
    public int? GetTracksColor() => __baseBotInternals.CurrentTick.BotState.TracksColor;

    /// <inheritdoc/>
    public void SetTracksColor(string tracksColor) => __baseBotInternals.BotIntent.TracksColor = tracksColor;

    /// <inheritdoc/>
    public int? GetGunColor() => __baseBotInternals.CurrentTick.BotState.GunColor;

    /// <inheritdoc/>
    public void SetGunColor(string gunColor) => __baseBotInternals.BotIntent.GunColor = gunColor;

    /// <inheritdoc/>
    public double CalcMaxTurnRate(double speed) => Constants.MaxTurnRate - 0.75 * Math.Abs(speed);

    /// <inheritdoc/>
    public virtual double CalcBulletSpeed(double firepower) => 20 - 3 * firepower;

    /// <inheritdoc/>
    public virtual double CalcGunHeat(double firepower) => 1 + (firepower / 5);

    /// <inheritdoc/>
    public virtual double CalcBearing(double direction) => NormalizeRelativeAngle(direction - Direction);

    /// <inheritdoc/>
    public virtual double CalcGunBearing(double direction) => NormalizeRelativeAngle(direction - GunDirection);

    /// <inheritdoc/>
    public virtual double CalcRadarBearing(double direction) => NormalizeRelativeAngle(direction - RadarDirection);

    /// <inheritdoc/>
    public virtual double DirectionTo(double x, double y) => NormalizeAbsoluteAngle(180 * Math.Atan2(y - Y, x - X) / Math.PI);

    /// <inheritdoc/>
    public virtual double BearingTo(double x, double y) => NormalizeRelativeAngle(DirectionTo(x, y) - Direction);

    /// <inheritdoc/>
    public virtual double GunBearingTo(double x, double y) => NormalizeRelativeAngle(DirectionTo(x, y) - GunDirection);

    /// <inheritdoc/>
    public virtual double RadarBearingTo(double x, double y) => NormalizeRelativeAngle(DirectionTo(x, y) - RadarDirection);

    /// <inheritdoc/>
    public virtual double DistanceTo(double x, double y)
    {
      var dx = x - X;
      var dy = y - Y;
      return Math.Sqrt(dx * dx + dy * dy);
    }

    /// <inheritdoc/>
    public virtual double NormalizeAbsoluteAngle(double angle) => (angle %= 360) >= 0 ? angle : (angle + 360);

    /// <inheritdoc/>
    public virtual double NormalizeRelativeAngle(double angle) => (angle %= 360) >= 0 ?
        ((angle < 180) ? angle : (angle - 360)) :
        ((angle >= -180) ? angle : (angle + 360));

    /// <inheritdoc/>
    public virtual double CalcDeltaAngle(double targetAngle, double sourceAngle)
    {
      double angle = targetAngle - sourceAngle;
      angle += (angle > 180) ? -360 : (angle < -180) ? 360 : 0;
      return angle;
    }

    /// <inheritdoc/>
    public virtual void OnConnected(ConnectedEvent connectedEvent) => Console.WriteLine($"Connected to {connectedEvent.ServerUri}");

    /// <inheritdoc/>
    public virtual void OnDisconnected(DisconnectedEvent disconnectedEvent)
    {
        var msg = $"Disconnected from {disconnectedEvent.ServerUri}";
        if (disconnectedEvent.StatusCode != null) {
          msg += $", status code: {disconnectedEvent.StatusCode}";
        }
        if (disconnectedEvent.Reason != null) {
          msg += $", reason: {disconnectedEvent.Reason}";
        }
        Console.WriteLine(msg);
    }

    /// <inheritdoc/>
    public virtual void OnConnectionError(ConnectionErrorEvent connectionErrorEvent) => Console.Error.WriteLine($"Connection error with {connectionErrorEvent.ServerUri}: " + connectionErrorEvent.Exception.Message);

    /// <inheritdoc/>
    public virtual void OnGameStarted(GameStartedEvent gameStatedEvent) { }

    /// <inheritdoc/>
    public virtual void OnGameEnded(GameEndedEvent gameEndedEvent) { }

    /// <inheritdoc/>
    public virtual void OnRoundStarted(RoundStartedEvent roundStatedEvent) { }

    /// <inheritdoc/>
    public virtual void OnRoundEnded(RoundEndedEvent roundEndedEvent) { }

    /// <inheritdoc/>
    public virtual void OnTick(TickEvent tickEvent) { }

    /// <inheritdoc/>
    public virtual void OnBotDeath(DeathEvent botDeathEvent) { }

    /// <inheritdoc/>
    public virtual void OnDeath(DeathEvent botDeathEvent) { }

    /// <inheritdoc/>
    public virtual void OnHitBot(HitBotEvent botHitBotEvent) { }

    /// <inheritdoc/>
    public virtual void OnHitWall(HitWallEvent botHitWallEvent) { }

    /// <inheritdoc/>
    public virtual void OnBulletFired(BulletFiredEvent bulletFiredEvent) { }

    /// <inheritdoc/>
    public virtual void OnHitByBullet(BulletHitBotEvent bulletHitBotEvent) { }

    /// <inheritdoc/>
    public virtual void OnBulletHit(BulletHitBotEvent bulletHitBotEvent) { }

    /// <inheritdoc/>
    public virtual void OnBulletHitBullet(BulletHitBulletEvent bulletHitBulletEvent) { }

    /// <inheritdoc/>
    public virtual void OnBulletHitWall(BulletHitWallEvent bulletHitWallEvent) { }

    /// <inheritdoc/>
    public virtual void OnScannedBot(ScannedBotEvent scannedBotEvent) { }

    /// <inheritdoc/>
    public virtual void OnSkippedTurn(SkippedTurnEvent skippedTurnEvent) { }

    /// <inheritdoc/>
    public virtual void OnWonRound(WonRoundEvent wonRoundEvent) { }

    /// <inheritdoc/>
    public virtual void OnCustomEvent(CustomEvent customEvent) { }
  }
}

File: bot-api/typescript/src/BaseBot.ts
import { IBaseBot } from "./IBaseBot.js";
import { BotInfo } from "./BotInfo.js";
import { BulletState } from "./BulletState.js";
import { Color } from "./graphics/Color.js";
import { IGraphics } from "./graphics/IGraphics.js";
import { BotEvent } from "./events/BotEvent.js";
import { Condition } from "./events/Condition.js";
import { ConnectedEvent } from "./events/ConnectedEvent.js";
import { DisconnectedEvent } from "./events/DisconnectedEvent.js";
import { ConnectionErrorEvent } from "./events/ConnectionErrorEvent.js";
import { GameStartedEvent } from "./events/GameStartedEvent.js";
import { GameEndedEvent } from "./events/GameEndedEvent.js";
import { RoundStartedEvent } from "./events/RoundStartedEvent.js";
import { RoundEndedEvent } from "./events/RoundEndedEvent.js";
import { TickEvent } from "./events/TickEvent.js";
import { BotDeathEvent } from "./events/BotDeathEvent.js";
import { DeathEvent } from "./events/DeathEvent.js";
import { HitBotEvent } from "./events/HitBotEvent.js";
import { HitWallEvent } from "./events/HitWallEvent.js";
import { BulletFiredEvent } from "./events/BulletFiredEvent.js";
import { HitByBulletEvent } from "./events/HitByBulletEvent.js";
import { BulletHitBotEvent } from "./events/BulletHitBotEvent.js";
import { BulletHitBulletEvent } from "./events/BulletHitBulletEvent.js";
import { BulletHitWallEvent } from "./events/BulletHitWallEvent.js";
import { ScannedBotEvent } from "./events/ScannedBotEvent.js";
import { SkippedTurnEvent } from "./events/SkippedTurnEvent.js";
import { WonRoundEvent } from "./events/WonRoundEvent.js";
import { CustomEvent } from "./events/CustomEvent.js";
import { TeamMessageEvent } from "./events/TeamMessageEvent.js";
import { BaseBotInternals } from "./internal/BaseBotInternals.js";
import { EnvVars } from "./EnvVars.js";
import { detectRuntime } from "./runtime/index.js";
import { Constants } from "./Constants.js";
import { MathUtil } from "./util/MathUtil.js";

/**
 * Abstract base class for bots implementing the IBaseBot interface.
 * Provides all state accessors, rate setters, color management, and math utilities.
 * Subclasses should override the event handler methods to implement bot behavior.
 */
export abstract class BaseBot implements IBaseBot {
  readonly TEAM_MESSAGE_MAX_SIZE = Constants.TEAM_MESSAGE_MAX_SIZE;
  readonly MAX_NUMBER_OF_TEAM_MESSAGES_PER_TURN = Constants.MAX_NUMBER_OF_TEAM_MESSAGES_PER_TURN;

  /** @internal */
  readonly _internals: BaseBotInternals;

  /**
   * Creates a new BaseBot with auto-detected configuration from environment variables.
   */
  constructor();
  /**
   * Creates a new BaseBot with the given bot info.
   * @param botInfo the bot info.
   */
  constructor(botInfo: BotInfo);
  /**
   * Creates a new BaseBot with the given bot info and server URL.
   * @param botInfo the bot info.
   * @param serverUrl the server URL.
   */
  constructor(botInfo: BotInfo, serverUrl: string);
  /**
   * Creates a new BaseBot with the given bot info, server URL, and server secret.
   * @param botInfo the bot info.
   * @param serverUrl the server URL.
   * @param serverSecret the server secret.
   */
  constructor(botInfo: BotInfo, serverUrl: string, serverSecret: string);
  constructor(botInfo?: BotInfo, serverUrl?: string, serverSecret?: string) {
    const info = botInfo ?? BaseBot._loadBotInfo(new.target.name);
    this._internals = new BaseBotInternals(
      this,
      info,
      serverUrl ?? null,
      serverSecret,
    );
  }

  private static _loadBotInfo(className: string): BotInfo {
    const runtime = detectRuntime();
    // Try <ClassName>.json in the current directory first (matches Python/Java behaviour)
    const text = runtime.readFile(`./${className}.json`);
    if (text) {
      try {
        return BotInfo.fromJson(text);
      } catch {
        // Fall through to env vars
      }
    }
    return new EnvVars(runtime).getBotInfo();
  }

  start(): void {
    this._internals.start();
  }

  go(): void {
    const currentTick = this._internals.getCurrentTickOrNull();
    const capturedTurnNumber = currentTick?.turnNumber ?? -1;
    if (currentTick != null) {
      this._internals.dispatchEvents(capturedTurnNumber);
    }
    // Pass captured turn number to execute() so it uses the same tick we dispatched events for.
    // Without this, execute() re-reads the live tickEvent which may have been updated by the
    // message handler between dispatchEvents() and execute(), causing skipped turns.
    this._internals.execute(capturedTurnNumber);
  }

  getMyId(): number { return this._internals.getMyId(); }
  getVariant(): string { return this._internals.getVariant(); }
  getVersion(): string { return this._internals.getVersion(); }
  getGameType(): string { return this._internals.getGameType(); }
  getArenaWidth(): number { return this._internals.getArenaWidth(); }
  getArenaHeight(): number { return this._internals.getArenaHeight(); }
  getNumberOfRounds(): number { return this._internals.getNumberOfRounds(); }
  getGunCoolingRate(): number { return this._internals.getGunCoolingRate(); }
  getMaxInactivityTurns(): number { return this._internals.getMaxInactivityTurns(); }
  getTurnTimeout(): number { return this._internals.getTurnTimeout(); }
  getTimeLeft(): number { return this._internals.getTimeLeft(); }
  getRoundNumber(): number { return this._internals.getRoundNumber(); }
  getTurnNumber(): number { return this._internals.getTurnNumber(); }
  getEnemyCount(): number { return this._internals.getEnemyCount(); }
  getEnergy(): number { return this._internals.getEnergy(); }
  isDisabled(): boolean { return this._internals.isDisabled(); }
  getX(): number { return this._internals.getX(); }
  getY(): number { return this._internals.getY(); }
  getDirection(): number { return this._internals.getDirection(); }
  getGunDirection(): number { return this._internals.getGunDirection(); }
  getRadarDirection(): number { return this._internals.getRadarDirection(); }
  getSpeed(): number { return this._internals.getSpeed(); }
  getGunHeat(): number { return this._internals.getGunHeat(); }
  getBulletStates(): ReadonlySet<BulletState> { return this._internals.getBulletStates(); }
  getEvents(): BotEvent[] { return [...this._internals.getEvents()]; }
  clearEvents(): void { this._internals.clearEvents(); }

  getTurnRate(): number { return this._internals.getTurnRate(); }
  setTurnRate(turnRate: number): void { this._internals.setTurnRate(turnRate); }
  getMaxTurnRate(): number { return this._internals.getMaxTurnRate(); }
  setMaxTurnRate(maxTurnRate: number): void { this._internals.setMaxTurnRate(maxTurnRate); }
  getGunTurnRate(): number { return this._internals.getGunTurnRate(); }
  setGunTurnRate(gunTurnRate: number): void { this._internals.setGunTurnRate(gunTurnRate); }
  getMaxGunTurnRate(): number { return this._internals.getMaxGunTurnRate(); }
  setMaxGunTurnRate(maxGunTurnRate: number): void { this._internals.setMaxGunTurnRate(maxGunTurnRate); }
  getRadarTurnRate(): number { return this._internals.getRadarTurnRate(); }
  setRadarTurnRate(radarTurnRate: number): void { this._internals.setRadarTurnRate(radarTurnRate); }
  getMaxRadarTurnRate(): number { return this._internals.getMaxRadarTurnRate(); }
  setMaxRadarTurnRate(maxRadarTurnRate: number): void { this._internals.setMaxRadarTurnRate(maxRadarTurnRate); }
  getTargetSpeed(): number { return this._internals.getTargetSpeed(); }
  setTargetSpeed(targetSpeed: number): void { this._internals.setTargetSpeed(targetSpeed); }
  getMaxSpeed(): number { return this._internals.getMaxSpeed(); }
  setMaxSpeed(maxSpeed: number): void { this._internals.setMaxSpeed(maxSpeed); }

  setFire(firepower: number): boolean { return this._internals.setFire(firepower); }
  getFirepower(): number { return this._internals.getFirepower(); }
  setRescan(): void { this._internals.setRescan(); }
  setFireAssist(enable: boolean): void { this._internals.setFireAssist(enable); }
  setInterruptible(interruptible: boolean): void { this._internals.setInterruptible(interruptible); }

  setAdjustGunForBodyTurn(adjust: boolean): void { this._internals.setAdjustGunForBodyTurn(adjust); }
  isAdjustGunForBodyTurn(): boolean { return this._internals.isAdjustGunForBodyTurn(); }
  setAdjustRadarForBodyTurn(adjust: boolean): void { this._internals.setAdjustRadarForBodyTurn(adjust); }
  isAdjustRadarForBodyTurn(): boolean { return this._internals.isAdjustRadarForBodyTurn(); }
  setAdjustRadarForGunTurn(adjust: boolean): void { this._internals.setAdjustRadarForGunTurn(adjust); }
  isAdjustRadarForGunTurn(): boolean { return this._internals.isAdjustRadarForGunTurn(); }

  addCustomEvent(condition: Condition): boolean { return this._internals.addCustomEvent(condition); }
  removeCustomEvent(condition: Condition): boolean { return this._internals.removeCustomEvent(condition); }

  setStop(): void;
  setStop(overwrite: boolean): void;
  setStop(overwrite?: boolean): void { this._internals.setStop(overwrite); }
  setResume(): void { this._internals.setResume(); }
  isStopped(): boolean { return this._internals.isStopped_(); }

  getTeammateIds(): ReadonlySet<number> { return this._internals.getTeammateIds(); }
  isTeammate(botId: number): boolean { return this._internals.isTeammate(botId); }
  broadcastTeamMessage(message: unknown): void { this._internals.broadcastTeamMessage(message); }
  sendTeamMessage(teammateId: number, message: unknown): void { this._internals.sendTeamMessage(teammateId, message); }

  getBodyColor(): Color | null { return this._internals.getBodyColor(); }
  setBodyColor(color: Color | null): void { this._internals.setBodyColor(color); }
  getTurretColor(): Color | null { return this._internals.getTurretColor(); }
  setTurretColor(color: Color | null): void { this._internals.setTurretColor(color); }
  getRadarColor(): Color | null { return this._internals.getRadarColor(); }
  setRadarColor(color: Color | null): void { this._internals.setRadarColor(color); }
  getBulletColor(): Color | null { return this._internals.getBulletColor(); }
  setBulletColor(color: Color | null): void { this._internals.setBulletColor(color); }
  getScanColor(): Color | null { return this._internals.getScanColor(); }
  setScanColor(color: Color | null): void { this._internals.setScanColor(color); }
  getTracksColor(): Color | null { return this._internals.getTracksColor(); }
  setTracksColor(color: Color | null): void { this._internals.setTracksColor(color); }
  getGunColor(): Color | null { return this._internals.getGunColor(); }
  setGunColor(color: Color | null): void { this._internals.setGunColor(color); }

  calcMaxTurnRate(speed: number): number {
    return Constants.MAX_TURN_RATE - 0.75 * Math.abs(MathUtil.clamp(speed, -Constants.MAX_SPEED, Constants.MAX_SPEED));
  }

  calcBulletSpeed(firepower: number): number {
    return 20 - 3 * MathUtil.clamp(firepower, Constants.MIN_FIREPOWER, Constants.MAX_FIREPOWER);
  }

  calcGunHeat(firepower: number): number {
    return 1 + MathUtil.clamp(firepower, Constants.MIN_FIREPOWER, Constants.MAX_FIREPOWER) / 5;
  }

  getEventPriority(eventType: string): number { return this._internals.getEventPriority(eventType); }
  setEventPriority(eventType: string, priority: number): void { this._internals.setEventPriority(eventType, priority); }

  isDebuggingEnabled(): boolean { return this._internals.isDebuggingEnabled(); }
  getGraphics(): IGraphics { return this._internals.getGraphics(); }

  // ---------------------------------------------------------------------------
  // Overridable event handler methods (no-op by default)
  // ---------------------------------------------------------------------------

  onConnected(_event: ConnectedEvent): void { /* override to handle */ }
  onDisconnected(_event: DisconnectedEvent): void { /* override to handle */ }
  onConnectionError(_event: ConnectionErrorEvent): void { /* override to handle */ }
  onGameStarted(_event: GameStartedEvent): void { /* override to handle */ }
  onGameEnded(_event: GameEndedEvent): void { /* override to handle */ }
  onRoundStarted(_event: RoundStartedEvent): void { /* override to handle */ }
  onRoundEnded(_event: RoundEndedEvent): void { /* override to handle */ }
  onTick(_event: TickEvent): void { /* override to handle */ }
  onBotDeath(_event: BotDeathEvent): void { /* override to handle */ }
  onDeath(_event: DeathEvent): void { /* override to handle */ }
  onHitBot(_event: HitBotEvent): void { /* override to handle */ }
  onHitWall(_event: HitWallEvent): void { /* override to handle */ }
  onBulletFired(_event: BulletFiredEvent): void { /* override to handle */ }
  onHitByBullet(_event: HitByBulletEvent): void { /* override to handle */ }
  onBulletHitBot(_event: BulletHitBotEvent): void { /* override to handle */ }
  onBulletHitBullet(_event: BulletHitBulletEvent): void { /* override to handle */ }
  onBulletHitWall(_event: BulletHitWallEvent): void { /* override to handle */ }
  onScannedBot(_event: ScannedBotEvent): void { /* override to handle */ }
  onSkippedTurn(_event: SkippedTurnEvent): void { /* override to handle */ }
  onWonRound(_event: WonRoundEvent): void { /* override to handle */ }
  onCustomEvent(_event: CustomEvent): void { /* override to handle */ }
  onTeamMessage(_event: TeamMessageEvent): void { /* override to handle */ }

  // ---------------------------------------------------------------------------
  // Calculation methods
  // ---------------------------------------------------------------------------

  calcBearing(direction: number): number {
    return this.normalizeRelativeAngle(direction - this.getDirection());
  }

  calcGunBearing(direction: number): number {
    return this.normalizeRelativeAngle(direction - this.getGunDirection());
  }

  calcRadarBearing(direction: number): number {
    return this.normalizeRelativeAngle(direction - this.getRadarDirection());
  }

  directionTo(x: number, y: number): number {
    return this.normalizeAbsoluteAngle(
      (Math.atan2(y - this.getY(), x - this.getX()) * 180) / Math.PI,
    );
  }

  bearingTo(x: number, y: number): number {
    return this.normalizeRelativeAngle(this.directionTo(x, y) - this.getDirection());
  }

  gunBearingTo(x: number, y: number): number {
    return this.normalizeRelativeAngle(this.directionTo(x, y) - this.getGunDirection());
  }

  radarBearingTo(x: number, y: number): number {
    return this.normalizeRelativeAngle(this.directionTo(x, y) - this.getRadarDirection());
  }

  distanceTo(x: number, y: number): number {
    const dx = x - this.getX();
    const dy = y - this.getY();
    return Math.sqrt(dx * dx + dy * dy);
  }

  normalizeAbsoluteAngle(angle: number): number {
    return ((angle % 360) + 360) % 360;
  }

  normalizeRelativeAngle(angle: number): number {
    let a = ((angle % 360) + 360) % 360;
    if (a > 180) a -= 360;
    return a;
  }

  calcDeltaAngle(targetAngle: number, sourceAngle: number): number {
    let angle = targetAngle - sourceAngle;
    angle += angle > 180 ? -360 : angle < -180 ? 360 : 0;
    return angle;
  }
}
